nx/lib/nx/binary_backend.ex

```diff
@@ -1687,23 +1687,56 @@ defmodule Nx.BinaryBackend do
   defp bin_slice(data, shape, size, start_indices, lengths, strides, output_shape) do
     start_indices = clamp_indices(start_indices, shape, lengths)
 
-    if top_dimension_slice?(tuple_size(shape), shape, output_shape) do
-      length = Nx.size(output_shape) * div(size, 8)
-      offset = div(length, elem(output_shape, 0)) * hd(start_indices)
-      binary_part(data, offset, length)
-    else
-      # Anchored around the start indices
-      weighted_shape = weighted_shape(shape, size, output_shape)
-      offset = weighted_offset(weighted_shape, start_indices)
-
-      # The weighted shape is altered such that we traverse
-      # with respect to the stride for each dimension
-      weighted_shape =
-        Enum.zip_with(weighted_shape, strides, fn {d, dim_size}, s ->
-          {d, dim_size + (s - 1) * dim_size}
-        end)
-
-      IO.iodata_to_binary(weighted_traverse(weighted_shape, data, size, offset))
+    is_top_dim_slice = top_dimension_slice?(tuple_size(shape), shape, output_shape)
+
+    cond do
```

Comment:
Turns out we can also just revert the top-level optimization as well. The `else`-block code works properly for all axes.
Another possibility is that we can apply this optimization only if the top-level stride is 1, as the first clause below.

I'm leaning towards this second possibility, because the "optimized" code for the strided case doesn't seem optimized at all

Reply:
Yeah the strided case makes things difficult, I'm okay with getting rid of the optimization in favor of the simpler general case. WDYT @josevalim ?

Reply:
Ditching the optimization sounds good to me. It is your call!

Reply:
Actually, let's go with the idea of only applying the optimization if the top level stride is 1. We likely want to keep this fast anyway because of the access syntax.